src/appleseed/renderer/modeling/camera/camera.cpp

```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
+        {
+            return curve.evaluate(param) - value;
+        };
+
+    const auto d = [&curve](float param)
+        {
+            return curve.evaluate_derivative(param);
+        };
+
+    bool success = foundation::find_root_newton(
+        f,
+        d,
+        a,
+        b,
+        eps,
+        10,
+        root
+    );
+
+    if (success)
+    {
+        // make sure neuton method didn't push root slightly out of bounds
+        root = foundation::clamp(root, a, max_param);
+
+        return foundation::evaluate_bezier3(t0, t1, t2, t3, root);
+    }
+
     //
-    // Shutter curves. Used in Camera::map_to_shutter_curve().
+    // Fallback to bisection.
     //
 
-    // Integrate and inverse opening/closing curve y = a * x + b to transform samples.
-    float map_sample_to_linear_curve(const float a, const float b, const float x)
+    success = foundation::find_root_bisection(
+        f,
+        a,
+        b,
+        eps,
+        21, // This would be enough to get 10e-6 accuracy for the whole [a,b] interval.
+        root);
+
+    if (!success)
     {
-        float constant;
-        if (a > 0.0f)
-        {
-            // Shutter is opening.
-            constant = 0.0f;
-        }
-        else
-        {
-            // Shutter is closing.
-            constant = 1.0f - (a / 2.0f) - b;
-        }
-        return (sqrt(2.0f * a * (x - constant) + b * b) - b) / a;
+        assert(false && "Cannot inverse sample shutter curve CDF");
+        RENDERER_LOG_ERROR("Cannot inverse sample shutter curve CDF of camera \"%s\"", camera.get_path().c_str());
     }
+
+    return foundation::evaluate_bezier3(t0, t1, t2, t3, root);
 }
 
-float Camera::map_to_shutter_curve(const float sample) const
+} // namespace
+
+// Linear curve consists of multiple linear patches
+float Camera::map_to_shutter_curve_impl_linear(const float sample) const
 {
-    if (0.0f <= sample && sample < m_inverse_cdf_open_point)
+    if (sample < m_inverse_cdf_open_point)
     {
         // Shutter is opening.
-        return map_sample_to_linear_curve(m_open_linear_curve_slope, 0.0f, sample);
+        return sqrt(m_shutter_curve_linear_open_multiplier * sample);
     }
-    else if (1.0f >= sample && sample > m_inverse_cdf_close_point)
+    else if (sample > m_inverse_cdf_close_point)
     {
         // Shutter is closing.
-        return map_sample_to_linear_curve(m_close_linear_curve_slope, -m_close_linear_curve_slope, sample);
+        return 1.0f - sqrt(m_shutter_curve_linear_close_multiplier * (sample - 1.0f));
     }
     else
     {
         // Shutter is fully opened.
-        return sample / m_shutter_pdf_max_height + m_normalized_open_end_time_half;
+        return m_shutter_curve_linear_fully_opened_multiplier * sample + m_shutter_open_end_time / 2.0f;
+    }
+}
+
+// Nonlinear curve consists of two bezier curves and constant patch between them
+float Camera::map_to_shutter_curve_impl_nonlinear(const float sample) const
```

Comment:
Maybe `_bezier` is clearer than `_nonlinear`?

Reply:
I wanted to name it _bezier, but finally we've got something that is not a Bezier curve itself, so nonlinear curve is more accurate name.
Should be discussed further though.

Reply:
It's not a Bézier curve if you consider the whole curve, but it's still a combination of linear and Bézier segments. If we were to add support for other curves (B-Spline?) then the `_nonlinear` suffix would be a problem.

Reply:
Let it be bezier then.